server/apps/app.go

```diff
@@ -32,46 +33,77 @@ type App struct {
 	BotUsername    string `json:"bot_username,omitempty"`
 	BotAccessToken string `json:"bot_access_token,omitempty"`
 
-	// Grants should be scopable in the future, per team, channel, post with regexp
+	// Grants should be scopable in the future, per team, channel, post with
+	// regexp.
 	GrantedPermissions Permissions `json:"granted_permissions,omitempty"`
+
+	// GrantedLocations contains the list of top locations that the
+	// application is allowed to bind to.
+	GrantedLocations Locations `json:"granted_locations,omitempty"`
 }
 
-type PermissionType string
-
-const (
-	PermissionUserJoinedChannelNotification = PermissionType("user_joined_channel_notification")
-	PermissionAddToPostMenu                 = PermissionType("add_to_post_menu")
-	PermissionAddGrants                     = PermissionType("add_grants")
-	PermissionActAsUser                     = PermissionType("act_as_user")
-	PermissionActAsBot                      = PermissionType("act_as_bot")
-)
-
-func (p PermissionType) Markdown() md.MD {
-	m := ""
-	switch p {
-	case PermissionAddToPostMenu:
-		m = "Add items to Post menu"
-	case PermissionUserJoinedChannelNotification:
-		m = "Be notified when users join channels"
-	case PermissionAddGrants:
-		m = "Add more grants (WITHOUT ADDITIONAL ADMIN CONSENT)"
-	case PermissionActAsUser:
-		m = "Use Mattermost REST API as connected users"
-	case PermissionActAsBot:
-		m = "Use Mattermost REST API as the app's bot user"
-	default:
-		m = "unknown permission: " + string(p)
+func (m *Manifest) ConfigMap() map[string]interface{} {
+	return map[string]interface{}{
+		"app_id":                string(m.AppID),
+		"display_name":          m.DisplayName,
+		"description":           m.Description,
+		"oauth2_callback_url":   m.OAuth2CallbackURL,
+		"homepage_url":          m.HomepageURL,
+		"root_url":              m.RootURL,
+		"requested_permissions": m.RequestedPermissions.toStringArray(),
+		"requested_locations":   m.RequestedLocations.toStringArray(),
 	}
-	return md.MD(m)
 }
 
-type Permissions []PermissionType
+func ManifestFromConfigMap(in interface{}) *Manifest {
+	m := &Manifest{}
+	c, _ := in.(map[string]interface{})
+	if len(c) == 0 {
+		return m
+	}
+	appID, _ := c["app_id"].(string)
+	m.AppID = AppID(appID)
+	m.DisplayName, _ = c["display_name"].(string)
+	m.Description, _ = c["description"].(string)
+	m.OAuth2CallbackURL = c["oauth2_callback_url"].(string)
+	m.HomepageURL = c["homepage_url"].(string)
+	m.RootURL = c["root_url"].(string)
+	m.RequestedPermissions = permissionsFromConfigArray(c["requested_permissions"])
+	m.RequestedLocations = locationsFromConfigArray(c["requested_locations"])
+	return m
+}
```

Comment:
@levb Any strong reason to not do this as something like:
```
b, _ := json.Marshal(in)
var m Manifest
_ := json.Unmarshal(b, &m)
return m
```
It is simpler at needs less maintenance (if we were to add or remove any field to the manifest, we have to change also this).

Reply:
A similar approach can be done for all the functions in this file.

Reply:
Another solution would be to create our own function using reflection to check the json tag and build the object, but why to reinvent the wheel?